### PR TITLE
upper bound for pyzmq/jupyter_client for notebook<7

### PR DESCRIPTION
**Pull request by mcg1969**
See this PR and linked issues:
https://github.com/jupyter/notebook/pull/6749

### CODE REVIEW

**Comment by mcg1969** [comment]
FYI, these pins will be built into notebook 6.5.5 already. But we need these for older versions, otherwise conda is likely to downgrade notebook to get newer versions of jupyter_client and pyzmq.

**Comment by mcg1969** [comment]
Notebook 5 already has an upper bound for `jupyter_client` but not for `pyzmq`. Talking with @rrosio it is my view that applying the `pymzq` pin is still worthwhile in that case.

**Review by markan — APPROVED** [review]
Looks reasonable to me.

**Review by Jrice1317 — APPROVED** [review]
LGTM

File: main.py [hunk 1]
```diff
@@ -1027,8 +1027,13 @@ def patch_record_in_place(fn, record, subdir):
 
     # notebook <5.7.6 will not work with tornado 6, see:
     # https://github.com/jupyter/notebook/issues/4439
+    # notebook <7 will not work with pyzmq>=25 and jupyter_client>=8, see:
+    # https://github.com/jupyter/notebook/pull/6749
     if name == "notebook":
         replace_dep(depends, "tornado >=4", "tornado >=4,<6")
+        if int(version.split('.', 1)[0]) < 7:
+            replace_dep(depends, "pyzmq >=17", "pyzmq >=17,<25")
+            replace_dep(depends, "jupyter_client >=5.3.4", "jupyter_client >=5.3.4,<8")
```
Comment (markan) [inline main.py:1036]: It looks like notebook version 5.0.0-6.0.1 ether pins to jupyter_client>=5.2.0 or doesn't pin at all. Do those need the upper bound as well? Or is some other constraint protecting us here?

Comment (markan) [reply main.py:1036]: In practice most of these are built for old versions of python, but notebook-5.7.9-py38_0 might still slip through.

Comment (mcg1969) [reply main.py:1036]: I'll add that

**Comment by lorepirri** [comment]
I've deleted the branch and change the settings of this repo to do it automatically after merge

File: main.py [hunk 2]
```diff
@@ -1027,8 +1027,15 @@ def patch_record_in_place(fn, record, subdir):
...
+            replace_dep(depends, "jupyter_client >=5.2.0", "jupyter_client >=5.2.0,<8")
+            replace_dep(depends, "jupyter_client", "jupyter_client <8")
```
Comment (mcg1969) [inline main.py:1039]: pins added in response to comment

**Pull request merged**